File: June-LeetCoding-Challenge-2020/Week2/35-search-insertion-position.py
class Solution:
    # Sol 1. Two pointers
    def searchInsert(self, nums: List[int], target: int) -> int:
        if len(nums) == 0:
            return 0
        left_bound = 0
        right_bound = len(nums) -1
        ptr = 0
        while left_bound <= right_bound:
            ptr = (left_bound+right_bound)//2
            if nums[ptr] == target:
                return ptr
            elif nums[ptr] < target:
                left_bound = ptr + 1
            else:
                right_bound = ptr - 1
        return left_bound 
    # TC: O(logn),n is the length of nums.
    # SC: O(1).

File: June-LeetCoding-Challenge-2020/Week2/35-search-insertion-position.cpp
// Time Complexity: O(logn) binary search tree
// Space Complexity: O(1)
// Runtime: 8 ms, faster than 71.15%
// Memory Usage: 9.7 MB, less than 41.13%
class Solution {
public:
    int getMiddle(int lower, int upper){
        return lower + (upper-lower)/2;
    }
    
    int searchInsert(vector<int>& nums, int target) {
        int lower, upper, mid;
        lower = 0;
        upper = nums.size()-1;
        
        
        while(lower <= upper){
            mid = getMiddle(lower, upper);
            
            if(target <= nums[mid]){
                upper = mid - 1;
            }else{
                lower = mid + 1;
            }        
        }
        
        return lower;
    }
};